### PR TITLE
🏗️ Scoped Query Selector: linting and audit

### PR DESCRIPTION
**Pull request by jridgewell**
This implements a much more rigorous `#querySelector` and `scopedQuerySelector` linting rule. We now reliably detect buggy `#querySelector` uses, and warn for non-escaped values in the selector.

This also implements the reverse, warning when we use `scopedQuerySelector` where it's not needed. This is nice because it avoids escaping hazards, and it's faster at runtime. 😃

Oh, and I made `escapeCssSelectorIdent` (who thought of this name?) much smaller. Still spec compliant!

### CODE REVIEW

File: third_party/css-escape/css-escape.js [hunk 1]
```diff
@@ -1,85 +1,26 @@
 /*! https://mths.be/cssescape v1.5.1 by @mathias | MIT license */
 
 
+var regex = /(\0)|^(-)$|([\x01-\x1f\x7f]|^-?[0-9])|([\x80-\uffff0-9a-zA-Z_-]+)|[^]/g;
```
Comment (rsimha-amp) [inline third_party/css-escape/css-escape.js]: Add a comment that explains what this regex is doing.

File: builtins/.eslintrc [hunk 2]
```diff
@@ -0,0 +1,6 @@
+{
+  "rules": {
+    "no-global": 2,
+    "query-selector": 2
```
Comment (rsimha-amp) [inline builtins/.eslintrc]: Why not add this to `ampproject/amphtml/.eslintrc`? The previous regex based presubmit check was applied globally.

File: third_party/css-escape/css-escape.js (hunk 1 above)
Comment (jridgewell) [reply third_party/css-escape/css-escape.js]: Done.

File: builtins/.eslintrc (hunk 2 above)
Comment (jridgewell) [reply builtins/.eslintrc]: Ads mainly, I don't want them to be linted.

Comment (rsimha-amp) [reply builtins/.eslintrc]: In that case, you could add `"query-selector": 2` to `ampproject/amphtml/.eslintrc` and disable the rule with `"query-selector": 0` in just the ads folder.

File: build-system/eslint-rules/query-selector.js [hunk 3]
```diff
@@ -0,0 +1,174 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
```
Comment (erwinmombay) [inline build-system/eslint-rules/query-selector.js]: 2018

File: build-system/eslint-rules/query-selector.js [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+    // If it's not a querySelector(All) call, I don't care about it.
+    const {property} = callee;
+    if (property.type !== 'Identifier' ||
+        !property.name.startsWith('querySelector')) {
+      return;
+    }
+
+    if (property.leadingComments) {
+      const ok = property.leadingComments.some(comment => {
+        return comment.value === 'OK'
```
Comment (erwinmombay) [inline build-system/eslint-rules/query-selector.js]: missing ;

Comment (rsimha-amp) [reply build-system/eslint-rules/query-selector.js]: Hah! We don't lint the stuff in `build-system/eslint-rules` :)

Comment (rsimha-amp) [reply build-system/eslint-rules/query-selector.js]: We probably should, though.

**Comment by jridgewell** [comment]
@rsimha-amp any idea why lgtm is failing?

**Comment by rsimha-amp** [comment]
It's due to https://discuss.lgtm.com/t/analysis-failed-could-not-build-the-merge-commit-and-the-base-commit/860. Fixed in #13460. Rebase on `master` and you should be good.

**Comment by jridgewell** [comment]
Perfect, just need a approval review now. 😉

**Pull request merged**